docs/source/daml/intro/11_Testing.rst

```diff
@@ -0,0 +1,254 @@
+  [PartyDetails {party = 'party-3cca7cc5', displayName = Some "Alice", isLocal = True},PartyDetails {party = 'party-4b70184e', displayName = Some "Bob", isLocal = True},PartyDetails {party = 'party-6e5b60bf', displayName = Some "USD_Bank", isLocal = True},PartyDetails {party = 'party-3d836540', displayName = Some "EUR_Bank", isLocal = True}]
+  daml> let alice = (parties!!0).party
+  daml> let eurbank = (parties!!3).party
+
+Now we can run upgrade steps just like we would in Script:
+
+.. code-block:: shell
+
+  daml> let rel = Relationship with issuer=eurbank; owner=alice
+  daml> initiateUpgrade rel
+  daml> acceptUpgrade rel
+
+All of the script return ``()`` so the REPL doesn't show us anything here. To make sure something is happening, let's query the ledger at this point.
+
+
+.. code-block:: shell
+
+  daml> query @Upgrade alice
+  [(<contract-id>,Upgrade {issuer = 'party-3d836540', owner = 'party-3cca7cc5'})]
+
+There's an ``Upgrade`` contract as expected. Now Alice didn't actually have any EUR holdings so we should be able to skip the ``performUpgrade`` step. Let's check whether the upgrade is complete according to the check functions:
+
+.. code-block:: shell
+
+  daml> ownerCheckUpgradeComplete rel
+  True
+  daml> issuerCheckUpgradeComplete rel
+  True
+
+As expected, there's nothing to upgrade so let's interactively complete this upgrade.
+
+.. code-block:: shell
+
+  daml> confirmCompletion rel
+  daml> completeUpgrade rel
+
+  As you can see, DAML Repl is able to mix and match scripts and custom interactions seamlessly. 
+
+Debug, Trace, and Stacktraces
+-----------------------------
+
+The above demonstrates nicely how to test the happy path, but what if a function doesn't behave as you expected? DAML has two functions that allow you to do fine-grained printf debugging: ``debug`` and ``trace``. Both allow you to print something to StdOut if the code is reached. The difference between ``debug`` and ``trace`` is similar to the relationship between ``abort`` and ``error``:
+
+- ``debug : Text -> m ()`` maps a text to an Action that has the side-effect of printing to StdOut.
+- ``trace : Text -> a -> a`` prints to StdOut when the expression is evaluated. 
+
+.. code-block:: shell
+
+  daml> let a : Script () = debug "foo"
+  daml> let b : Script () = trace "bar" (debug "baz")
+  [Daml.Script:378]: "bar"
+  daml> a
+  [DA.Internal.Prelude:540]: "foo"
+  daml> b
+  [DA.Internal.Prelude:540]: "baz"
+  daml>
+
+If in doubt, use ``debug``. It's the easier of the two to interpret the results of.
+
+The thing in the square brackets is a stack trace. It'll tell you the DAML file and line number that triggered the printing, but often no more than that because full stacktraces could violate subtransaction privacy quite easily. If you want to enable stacktraces for some purely functional code in your modules, you can use the machinery in :doc:`/daml/stdlib/DA-Stack` to do so, but we won't cover that any further here.
+
+Contention
```

Comment:
This section feels a bit out of place. It isn’t about testing or debugging but a mix of explaining the different types of contentions and potential fixes. If you want to keep it in this section, maybe try to rephrase it in terms of starting with the errors you encounter and then explaining the reason for it and the fix.